62.CAD/Hatch.cpp

```diff
@@ -0,0 +1,753 @@
+		if (std::abs(a) > std::max(std::abs(b), cubCoeffMag) * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Quartic<double>::construct(a, b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 4);
+		}
+		else if (abs(b) > quadCoeffMag * QUARTIC_THRESHHOLD)
+		{
+			auto res = equations::Cubic<double>::construct(b, c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 3);
+		}
+		else
+		{
+			auto res = equations::Quadratic<double>::construct(c, d, e).computeRoots();
+			memcpy(&t[0], &res.x, sizeof(double) * 2);
+		}
+
+		// If either is NaN or both are equal
+		// Same as: 
+		// if (t[0] == t[1] || isnan(t[0]) || isnan(t[1]))
+		if (!(t[0] != t[1]))
+			t[0] = t[0] != t_start ? t_start : t_end;
+
+		//// TODO: check that this clamp works with t[i] as NaN
+		//for (auto i = 0; i < 2; i++)
+		//	t[i] = nbl::core::clamp(t[i], t_start, t_end);
+		//
+		//// fix up a fuckup where both roots are NaN or were on the same side of the valid integral
+		//
+		//// TODO: do some Halley or Householder method steps on t
+		////while ()
+		////{
+		////}
+
+		// neither t still not in range, your beziers don't intersect
+		return t;
+	}
+};
+
+Hatch::QuadraticBezier Hatch::splitCurveRange(const QuadraticBezier& bezier, double minT, double maxT)
+{
+	assert(maxT > minT);
+	assert(0.0 <= minT && minT <= 1.0);
+	assert(0.0 <= maxT && maxT <= 1.0);
+    return splitCurveTakeUpper(splitCurveTakeLower(bezier, maxT), minT / maxT);
+}
+
+Hatch::QuadraticBezier Hatch::Segment::getSplitCurve() const
+{
+    return splitCurveRange(*originalBezier, t_start, t_end);
+}
+
+bool Hatch::Segment::isStraightLineConstantMajor() const
+{
+	auto major = (uint)SelectedMajorAxis;
+	const double p0 = originalBezier->P0[major], 
+		p1 = originalBezier->P1[major], 
+		p2 = originalBezier->P2[major];
+	//assert(p0 <= p1 && p1 <= p2); (PRECISION ISSUES ARISE ONCE MORE)
+	return abs(p1 - p0) <= exp2(-24) && abs(p2 - p0) <= exp(-24);
+}
+
+// copied from curves.cpp
+//TODO: move this to cpp-compat hlsl builtins
+static float64_t2 LineLineIntersection(const float64_t2& p1, const float64_t2& v1, const float64_t2& p2, const float64_t2& v2)
+{
+	float64_t denominator = v1.y * v2.x - v1.x * v2.y;
+	float64_t2 diff = p1 - p2;
+	float64_t numerator = dot(float64_t2(v2.y, -v2.x), float64_t2(diff.x, diff.y));
+
+	if (abs(denominator) < 1e-15 && abs(numerator) < 1e-15)
+	{
+		// are parallel and the same
+		return (p1 + p2) / 2.0;
+	}
+
+	float64_t t = numerator / denominator;
+	float64_t2 intersectionPoint = p1 + t * v1;
+	return intersectionPoint;
+}
+
+std::array<double, 2> Hatch::Segment::intersect(const Segment& other) const
+{
+	auto major = (int)SelectedMajorAxis;
+	auto minor = 1-major;
+
+	std::array<double, 2> result = { core::nan<double>(), core::nan<double>() };
+	int resultIdx = 0;
+
+	// Use line intersections if one or both of the beziers are linear (a = 0)
+	const bool selfLinear = isLineSegment(*originalBezier),
```

Comment:
I don't like this multi-liner stuff you've been doing, code alignment and spacing is a bit fucked. just write const bool twice, you'll be fine